### ISSUE TITLE
Allow for custom resolvers

### ISSUE THREAD

**Issue by mihaeu** [PHP-DI/PHP-DI#553]
Hey there, 

first off: thaaaank you so much for all the work you put into this wonderful project :+1: :smile: 

We've started using PHP DI on a large scale application and so far we're really happy with it. However we have to often specify definitions for classes which need string inputs (like named loggers).

Digging through your code I saw that this would be possible by adding a customer `Resolver`. However the Invoker is not injectable in either the `ContainerBuilder` or the `Container` class.

Would you be interested in a PR which allows for customer resolvers? I'd be glad to help out.

Cheers,
Mike

**Comment by mnapoli** [comment]
Hi! Thanks for the feedback :) And for the idea.

I think this could indeed be doable since the resolver is public API and strongly "standardized" since it's a separate package (https://github.com/PHP-DI/Invoker).

Just to be sure, you mean that you want to implement that interface right: https://github.com/PHP-DI/Invoker/blob/master/src/ParameterResolver/ParameterResolver.php ?

Feel free to send a pull request. I think you will need to add new methods to the `ContainerBuilder` and you'll need to pass the objects into many constructors. I hope this will not complicate everything too much but we'll see.

**Comment by mihaeu** [comment]
Yeah, I saw the Invoker package. Three problem won't be the constructors. Right now the issue lies in Container::getInvoker because that is where the configuration is hard coded and that's where the customisation has to happen. So it's about injecting dependencies into the dependency injector 😀

A simple solution would be setter injection of the Invoker or the Resolver connection. But that would be ugly. I think the best is what you suggested as well: making sure the ContainerBuilder gets the means to pass an alternative Invoker to the container.

**Comment by kakenbok** [comment]
#693

**Comment by guilhermeaiolfi** [comment]
So, I'm here with the same problem. Can't we simply turn the 'getInvoker' public? Because injecting the invoker won't solve my problem. I need to prepend the TypeHintResolver class in that list of resolvers. But I want to keep this line:

```php
new DefinitionParameterResolver($this->definitionResolver),
```

That I can NOT do outside the container, because ```$this->definitionResolver``` is private there.

Getting the invoker I can then ```getParameterResolver()->prependResolver``` or ```->appendResolver```
